### PR TITLE
Support Erlang/OTP 23

### PR DESCRIPTION
**Pull request by janl**
this PR pulls together all the various sub-module changes required to support Erlang/OTP 23.

Closes https://github.com/apache/couchdb/issues/3115

This currently includes a commit from https://github.com/apache/couchdb/pull/3421 because I couldn't run this locally otherwise. The PRs can land one after another, though.

### CODE REVIEW

File: src/couch/include/couch_db.hrl [hunk 1]
```diff
@@ -219,3 +219,29 @@
...
+%     ?STACKTRACE(exit, badarg, ErrorStackTrace)
+%         % do stuff with ErrorStackTrace
+%         % ...
+% end,
+
+% for features specific to Erlang/OTP version 20.x (and later versions)
+-ifdef(OTP_MAJOR_VSN_20).
+-else.
+-define(ERLANG_OTP_VERSION_21_FEATURES, true).
+-endif.
```
Comment (nickva) [inline src/couch/include/couch_db.hrl:238]: There might be an easier way to detect 20 vs 21+ we already do it in fabric2 https://github.com/apache/couchdb/blob/00811b4a9c4df56defb7456431d44abb7c0baa79/src/fabric/src/fabric2_util.erl#L386-L387

File: rebar.config.script [hunk 2]
```diff
@@ -92,6 +92,9 @@ case VerList of
     _ -> ok
 end.
 
+[OptMajorVersion|_] = VerList.
+OtpMajorVersion20 = case OptMajorVersion of 20 -> "true"; _ -> "false" end.
+os:putenv("COUCHDB_OTP_MAJOR_VERSION_20", OtpMajorVersion20).
```
Comment (nickva) [inline rebar.config.script:97]: If we only need to check this in Erlang code, we can rely on `OTP_RELEASE` directly:

https://erlang.org/doc/reference_manual/macros.html#predefined-macros

https://github.com/apache/couchdb/blob/00811b4a9c4df56defb7456431d44abb7c0baa79/src/fabric/src/fabric2_util.erl#L386-L387

**Comment by iilyak** [comment]
Please test how couch_log application work. I tried couch_log on OTP 22 and it was not working correctly for the following test. 
1. Define new clause in any of the named gen_server. For example `couch_server`:
  ```
      handle_info(die, State) ->
          ok = nok,
          {noreply, State};          
  ```
2. Start CouchDB `dev/run --admin=adm:pass`
3. Start remsh session `dev/remsh`
4. Send die message `couch_server ! die`
5. Check logs. You should see a log message about couch_server crashing (something matching the following pattern `"gen_server ~w terminated with reason: ~s~n"`).

This is due to significant changes in OTP logging subsystem.

**Comment by janl** [comment]
@iilyak this is what I get in the log:

```
[error] 2021-03-15T16:30:38.242690Z node1@127.0.0.1 <0.278.0> -------- couch_server terminating with {{badmatch,nok},[{couch_server,handle_info,2,[{file,"src/couch_server.erl"},{line,690}]},{gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,689}]},{gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,765}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,226}]}]}, state {server,"/Users/jan/Work/asf/couchdb/dev/lib/node1/data",[{"couch",couch_bt_engine}],62,0,"Mon, 15 Mar 2021 16:18:04 GMT",false,redacted,couch_dbs_1,couch_dbs_pid_to_name_1,couch_dbs_locks_1}
```

**Comment by iilyak** [comment]
I've got the following using 20.3.8.26

```
[error] 2021-03-15T20:47:39.317559Z node1@127.0.0.1 <0.250.0> -------- couch_server terminating with {{badmatch,nok},[{couch_server,handle_info,2,[{file,"src/couch_server.erl"},{line,719}]},{gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,616}]},{gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,686}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}, state {server,"/Users/iilyak@ca.ibm.com/dev/db.mono/src/couchdb/dev/lib/node1/data",[{"couch",couch_bt_engine}],62,0,"Mon, 15 Mar 2021 20:46:27 GMT",false,redacted,couch_dbs_1,couch_dbs_pid_to_name_1,couch_dbs_locks_1}
[error] 2021-03-15T20:47:39.318413Z node1@127.0.0.1 <0.250.0> -------- gen_server couch_server_1 terminated with reason: no match of right hand value nok at couch_server:handle_info/2(line:719) <= gen_server:try_dispatch/4(line:616) <= gen_server:handle_msg/6(line:686) <= proc_lib:init_p_do_apply/3(line:247)
  last msg: die
     state: {server,"/Users/iilyak@ca.ibm.com/dev/db.mono/src/couchdb/dev/lib/node1/data",[{"couch",couch_bt_engine}],62,0,"Mon, 15 Mar 2021 20:46:27 GMT",false,{0,#Ref<0.2205975427.3573678086.33533>,#Ref<0.2205975427.3573678086.33534>},couch_dbs_1,couch_dbs_pid_to_name_1,couch_dbs_locks_1}
    extra: []
[error] 2021-03-15T20:47:39.323063Z node1@127.0.0.1 <0.250.0> -------- CRASH REPORT Process couch_server_1 (<0.250.0>) with 0 neighbors crashed with reason: no match of right hand value nok at couch_server:handle_info/2(line:719) <= gen_server:try_dispatch/4(line:616) <= gen_server:handle_msg/6(line:686) <= proc_lib:init_p_do_apply/3(line:247); initial_call: {couch_server,init,['Argument__1']}, ancestors: [couch_primary_services,couch_sup,<0.244.0>], message_queue_len: 0, messages: [], links: [<0.247.0>], dictionary: [], trap_exit: true, status: running, heap_size: 4185, stack_size: 27, reductions: 6727
```

As you can see some of the messages are missing in your output.

Also while trying this out I realized that my instructions are no longer correct after a merge of couch_server work. The `couch_server ! die` should be `couch_server_1 ! die` now.

**Comment by nickva** [comment]
The logging issue is probably there since 21 when the introduced a whole new logging infrastructure. I remember advising someone at Cloudant to use Erlang 20 for locally development because they would not "see" some of the gen_server crash reports when they ran with 21.

I am not sure how hard it would be to fix it, hoping it would just a matter of configuring the additional compatibility mode:

 https://erlang.org/doc/man/sasl_app.html#deprecated-error-logger-event-handlers-and-configuration
 
 https://erlang.org/doc/apps/kernel/logger_chapter.html#compatibility
 
 https://erlang.org/doc/man/kernel_app.html#logger_level

However since this issue has been there at least since 21, and we already shipped a few releases supporting that OTP version, it's probably not a hard blocker for 23?

**Comment by wohali** [comment]
Note that our binary packages still ship with 20.3.8.25.

This being broken would be a primary reason to continue to ship against that version until it's fixed (@rnewson was worried about this a while back, see dev@ for details.)

**Comment by nickva** [comment]
@wohali good point, agree

**Comment by janl** [comment]
I tried setting `logger_sasl_compatible` to `true` in our `sys.config`, but that either doesn't work (pointers welcome), or doesn’t do what we want. The error log is the same, missing the extra output.

So we are missing last message and gen_server stat info? I agree that it is less convenient than before, but I wouldn’t call this “broken”.

Especially if we keep our binary packages on 20 for the time being.

**Comment by janl** [comment]
same with adding this to `vm.args` h/t rnewson

**Comment by janl** [comment]
So what’s the verdict here, are we going with 23-support-is-good-enough-for now, or are we looking for more changes?

I’m happy to file the logging issue as a separate ticket

**Comment by nickva** [comment]
@janl agree, let's have a separate issue for logging, I don't think this should be a blocker for 23 support as it wasn't a blocker for 21 and 22.

**Comment by wohali** [comment]
+1, with the proviso I don't intend to move the binaries/Docker off of 20.3.8.* until logging is fixed.

**Comment by nickva** [comment]
While we're talking about it, this should fix the error_logger in OTP >= 21 https://github.com/apache/couchdb/pull/3433

**Comment by nickva** [comment]
@janl while porting this PR to `main` it looks like inadvertently swapped the `?OTP_RELEASE` macro name with `?OTP_VERSION`. 

Notice in https://github.com/apache/couchdb/pull/3422#discussion_r593935304 it's `?OTP_RELEASE` but the final PR had `?OTP_VERSION` in https://github.com/apache/couchdb/pull/3422/files#diff-f95fbb47ac3f26db5ba353a42e63b998a104084db1aa3c0d149179163fa777fbR237

To double check: https://erlang.org/doc/reference_manual/macros.html#predefined-macros

**Pull request merged**